engine/launcher/src/main/scala/org/enso/launcher/releases/EngineReleaseProvider.scala

```diff
@@ -0,0 +1,131 @@
+    */
+  def packageFileName: String = {
+    val os = OS.operatingSystem match {
+      case OS.Linux   => "linux"
+      case OS.MacOS   => "macos"
+      case OS.Windows => "windows"
+    }
+    val arch = OS.architecture
+    val extension = OS.operatingSystem match {
+      case OS.Linux   => ".tar.gz"
+      case OS.MacOS   => ".tar.gz"
+      case OS.Windows => ".zip"
+    }
+    s"enso-engine-$version-$os-$arch$extension"
+  }
+}
+
+/**
+  * Wraps a generic [[ReleaseProvider]] to provide engine releases from it.
+  */
+class EngineReleaseProvider(releaseProvider: ReleaseProvider) {
+  private val tagPrefix = "enso-"
+
+  /**
+    * Returns the version of the most recent engine release.
+    */
+  def findLatest(): Try[SemVer] =
+    releaseProvider.listReleases().flatMap { releases =>
+      val versions =
+        releases.map(_.tag.stripPrefix(tagPrefix)).flatMap(SemVer(_))
+      versions.sorted.lastOption.map(Success(_)).getOrElse {
+        Failure(ReleaseProviderException("No valid engine versions were found"))
+      }
+    }
+
+  /**
+    * Fetches release metadata for a given version.
+    */
+  def getRelease(version: SemVer): Try[EngineRelease] = {
+    val tag = tagPrefix + version.toString
+    for {
+      release <- releaseProvider.releaseForTag(tag)
+      manifestAsset <-
+        release.assets
+          .find(_.fileName == Manifest.DEFAULT_MANIFEST_NAME)
+          .toRight(
+            ReleaseProviderException(
+              s"${Manifest.DEFAULT_MANIFEST_NAME} file is mising from " +
+              s"release assets."
+            )
+          )
+          .toTry
+      manifestContent <- manifestAsset.fetchAsText().waitForResult()
+      manifest <-
+        Manifest
+          .fromYaml(manifestContent)
+          .toRight(
+            ReleaseProviderException("Cannot parse attached manifest file.")
+          )
+          .toTry
+    } yield EngineRelease(version, manifest, release)
+  }
+
+  /**
+    * Downloads the package associated with the given release into
+    * `destination`.
+    *
+    * @param release the release to download the package from
+    * @param destination name of the file that will be created to contain the
+    * downloaded package
+    */
+  def downloadPackage(
+    release: EngineRelease,
+    destination: Path
+  ): TaskProgress[Unit] = {
+    val packageName = release.packageFileName
+    release.release.assets
+      .find(_.fileName == packageName)
+      .map(_.downloadTo(destination))
+      .getOrElse {
+        TaskProgress.immediateFailure(
+          ReleaseProviderException(
+            s"Cannot find package `$packageName` in the release."
+          )
+        )
+      }
+  }
+}
+
+/**
+  * Default [[EngineReleaseProvider]] that uses the GitHub Release API.
+  */
+object EngineReleaseProvider
+    extends EngineReleaseProvider(
+      new GithubReleaseProvider(
+        "enso-org",
+        "enso-staging" // TODO [RW] move from staging to the main repository
```

Comment:
When is this `TODO` going to be actioned?

Reply:
I see a few reasonable possibilities:
- I could potentially change it with this PR (but then reviewers would have to change it if they want to test locally if the installation works)
- it could be after most launcher stuff is done (I guess after self-update mechanism is done as it will also need testing from staging)
- when the first official Enso release is released (because then the launcher can actually use the original repository, currently the original repository is empty so it makes the launcher unusable).

Which one do you think is best? (I will add to this comment when we plan to resolve it.)

Reply:
I would suggest doing it when we publish the first prerelease release, which I plan on doing once the launcher is done.